ES6 at Last

mramato opened the pull request:
This PR contains all of the actual code changes to make ES6 work. In order to make this change actually reviewable, the `es6-staging` branch has all of the boilerplate changes that affect every Sandcastle demo and requirejs module. That leaves only ~115 files left for review, many of which are trivial changes or renames.

Here's a high level summary of all the areas that changed:

### What does ES6 modules mean for internal Cesium development?

At a high level:

`requirejs` is gone, `define` is gone. You now import code via the [import](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import) directive. i.e. 

```js
import Cartesian3 from './Cartesian3.js';
````

Instead of returning your object at the end of the file, you use [export](https://developer.mozilla.org/en-US/docs/web/javascript/reference/statements/export). Cesium uses default exports for it's internal modules i.e. 

```js
export default Cartesian3;
```

`use strict` is gone, modules are always strict.

While we use ES6 modules, Cesium still forbids the use of ES6 features, such as "new Promise", async/await/etc..

Other good stuff: Obviously this means way less boilerplate code, but how much?  We've cut over ~25k lines of code for our code base!  No more giant define blocks at the top of the file, no more wasting 20 minutes because your variables were out of order either!  The size of Cesium's npm and release packages have also dropped by ~9MB, and I think we can get them even smaller.

### eslint 

There are a handful of new `.eslintrc.json` files, mostly to identify the files that are still AMD modules (Sandcastle/Workers). These are needed because you can't change the parser type with a comment directive (since the parser is the thing reading the file). We can finally detect unusued modules!  So those have all been cleaned up as well.

### requirejs -> rollup & clean-css

requirejs, almond, and karma-requirejs have all been removed.  We now use rollup for building and minifying (via uglify) JS code and clean-css for css.  These changes are fairly straight-forward and just involve calling rollup instead of requirejs in the build process.

Overall build time is significantly faster. CI is ~11 minutes compared to ~17 in master.  Running `makeZipFile` on my machine takes 69 seconds compared to 112 seconds in master. There's probably plenty of room for additional optimization here too.

We wrote an published a small npm module, rollup-plugin-strip-pragma, for stripping the requirejs pragmas we use out of the release builds. This is maintained in the `Tools/rollup-plugin-strip-pragma` directory.

As for what we produce. The built version of Cesium is now a UMD module. So it should work anywhere that hasn't made the jump to ES6 yet.  For users that were already using the "legacy" combined/minified approach, nothing changes.

One awesome thing about roll-up is that it compiles all of the workers at once and automatically detects shared codes and generates separate bundles under the hood.  This means the size of our worker modules shrink dramatically and Cesium itself will load them much faster. The total minified/gzipped size of all workers in master is 2.6 MB compared to 225 KB in this branch! This should be most noticeable on demos like Geometry & Appearances which load lots of workers for the various geometry typs.

roll-up is also used to build Cesium Viewer, which is now an ES6 app.

We use clean-css via gulp and it is also a straightforward change from requirejs that requires no special mention.

### Workers 

While the spec allows for ES6 Web Workers, no browser actually supports them yet. That means we needed a way to get our workers into non-ES6 form. Thankfully, roll-up can generate AMD modules, which means we now have a build step to compile our Worker source code back into AMD and use the existing TaskProcessor to load and execute them. This build step is part of the standard `build` task and is called `createWorkers`. During development, these "built" workers are un-optimized so you can still debug them and read the code.

Since there is a build step, that means if you are changing code that affects a worker, you need to re-run `build`, or you can use the `build-watch` task to do it automatically.

The ES6 versions of Worker code has moved into `Source/WorkersES6` and we build the workers into their "old home" of `Source/Workers`. `cesiumWorkerBootstrapper` and `transferTypedArrayTest` which were already non-AMD ES5 scripts remain living in the Workers directory.

Surprisingly little was changed about TaskProcessor or the worker system in general, especially considering that I thought this would be one of the major hurdles.

### ThirdParty

A lot of our ThirdParty either already had a hand-written wrapper for AMD (which I updated to ES6) or had UMD which created problems when importing the same code in both Node and the browser. I basically had to update the wrapper of every third-party library to fix these problems.  In some cases I updated the library version itself (Autolinker, topojson). Nothing to be too concerned about, but future clean-up would be using npm versions of these libraries and auto-generating the wrappers as needed so we don't hand-edit things.

### Sandcastle

Sandcastle is eternal and manages to live another day in it's ancient requirejs/dojo 1.x form.  Sandcastle now automatically uses the ES6 version of Cesium if it is available and fallsback to the ES5 unminified version if it is now.  The built version of Sandcastle always uses CesiumUnminified, just like master.  This means Sandcastle still works in IE11 if you run the combine step first (or use the relase zip)

* Removed Cesium usage from Sandcastle proper, since it wasn't really needed
* Generate a VERSION propertyin the gallery index since Cesium is no longer being included.
* Remove requirejs from Sandcastle bucket
* Update bucket to use the built version of Cesium if it is available by fallbackto the ES6 version during development.
* Standalone.html was also updated

There's a bit of room for further clean-up here, but I think this gets us into master.  I did not rename `bucket-requirejs.html` because I'm pretty sure it would break previously shared demos.  We can put in some backwards compatible code later on if we want. (But I'd rather just see a full Sandcastle rewrite).

### Specs

Specs are now all ES6, except for TestWorkers, which remain standard JS worker modules. This means you can no longer run the unbuilt unit tests in IE11. No changes for Chrome and Firefox.

Since the specs use ES6 modules and built Cesium is an ES5 UMD, I added a `build-specs` build step which generates a combined ES5 version of the specs which rely on Cesium as a global variable. We then inject these files into jasmine instead of the standard specs and everything works exactly as it did before.  SpecRunner.html has been updated to inject the correct version of the script depending on the build/release query parameters.

The Specs must always use Cesium by importing `Source/Cesium.js`, this is so we can replace it with the built Cesium as describe above. 

There's a bunch of room for clean-up here, such as unifying our two copies of jasmine into a single helper file, but I didn't want to start doing that clean-up as part of this already overly big PR.  The important thing is that we can still test the built version and still test on IE/Edge as needed.

I also found and fixed two bugs that were causing failing unit tests, one in BingMapsImageryProviderSpec.js (which was overwriting `createImage` andnot setting it back) and ShadowVolumeAppearance.js (which had a module level caching bug). I think these may have been the cause of random CI failures in master as well, but only time will tell.

For coverage, we had to switch to `karma-coverage-istanbul-instrumenter` for native ES6 support, but that's it.

Finally, I updated appveryor to build Cesium and run the built tests under IE.  We still don't fail the build for IE, but we should probably fix that if we want to keep it going.

### NodeJS

When NODE_ENV is `production`, we now require in the minified CesiumJS directly, which works great because it's now a UMD module.  Otherwise, we use the excellant `esm`package to load individual modules, it was a fairly straightforward swap from our old requirejs usage.  We could probably drop esm too if we don't care about debugging or if we provie source maps at some point.

### What's next?

I tried to change as little formatting as possible to make a code review possible, however we now have uneeded indentation throughout our code. We should look into doing a code-wide autoformat, probably using eslint rules or one of the standard tools that work with it (this would also make it easy to maintain consistent formatting going forward). CC #5369

We should now be able to provide source maps for our users to make debugging with the minified JS code possible (not to mention make our lives easier). I'll write up an issue

Should update https://cesium.com/docs/tutorials/cesium-and-webpack/ or perhaps it is so trivial it's no longer needed.

Allow ES6/7 features: We should be able to keep our minimal build step policy and still use newer features like async/await/Promise/etc.. This is because Chrome/Firefox/Safari all support these features natively.  We would then only run the final output through Babel as part of the combine/minifyRelease processes. This would make our code a lot nicer with almost no overhead during day to day development (but more research is needed).

I need to review our developer documentation and make any minor updates so it's not wrong.  Ultimately, I realized while doing this that we need better infrastructure documentation over all.

And of course, lots of clean-up possible, I mentioned some of it above but I'll write separate issues as I go through my notes.

I'm sure I left out some details and I'm sure there were some next steps that I meant to write down and have since forgot, but I think the above should give everyone a fighting chance to review this PR.

I'm going to continue self-reviewing everything but this is ready for final review from whomever is wiling to take the time to look.

Fixes #8122

cesium-concierge commented:
Thanks for the pull request @mramato!
* :heavy_check_mark: Signed CLA found.
* :grey_question: Changes to third party files were made.
    * Looks like a file in one of our ThirdParty folders (`ThirdParty/, Source/ThirdParty/`) has been added or modified. Please verify that it has a section in [LICENSE.md](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/LICENSE.md) and that its license information is up to date with this new version.
* :grey_question: Unit tests were not updated.
    * Make sure you've [updated tests](https://github.com/AnalyticalGraphicsInc/cesium/tree/master/Documentation/Contributors/TestingGuide) to reflect your changes, added tests for any new code, and ran the code coverage tool.

Reviewers, don't forget to make sure that:

- [ ] Cesium Viewer works.
- [ ] Works in 2D/CV.
- [ ] Works (or fails gracefully) in IE11.

mramato commented:
For anyone that doesn't clone our repository but wants to try this out:

pre-release zip: http://cesium-dev.s3-website-us-east-1.amazonaws.com/cesium/es6-at-last/Cesium-1.61.0-es6-at-last32318.zip

npm package: http://cesium-dev.s3-website-us-east-1.amazonaws.com/cesium/es6-at-last/cesium-1.61.0-es6-at-last32318.tgz

`Source/Cesium.js` is the ES6 entry point you'll want to import

TJKoury commented:
@mramato Use [js-beautify](https://github.com/beautify-web/js-beautify) in the gulp task 'convertToModules' to clean up the indent issues.

markerikson commented:
or, say, Prettier :)

mramato commented:
Leaving the indenting for now is intentional.  `Prettier` is one of the tools we want to look at. (will also add js-beutify to the list)  But we definitely want to do it as a separate PR with no other changes than formatting, otherwise it makes stuff hard to review, so it won't be part of the initial submit.

We also have some special formatting needs because stuff like 4x4 matrices in our code are formatted for readability and we want to be able to maintain that. A blanket reformatting is going to introduce other messiness as well we'll want to clean up.

markerikson commented:
Yeah, and I totally agree with doing that later.

FWIW, here's Prettier's docs on ignoring files and chunks of code:

https://prettier.io/docs/en/ignore.html

puckey commented:
I noticed classes are being exported with `export default ExportedName`. Using `export const ExportedName = ...` instead would allow for IDE features like VSCode's intelliSense and WebStorm to do autocompletion while typing.. A quick search came up with this article with other reasons to stop using `export default`: https://humanwhocodes.com/blog/2019/01/stop-using-default-exports-javascript-module/

thw0rted commented:
I have some thoughts, so I'll try to capture them as I read through the notes:

* First and foremost: this is awesome, it must have been a ton of work but I'm glad you all did it!
* @puckey covered this but I don't see a lot of `export default` anymore.
* Have you considered distributing a built `esm.js` artifact?  As an ES6 consumer of Cesium, I can either consume the UMD artifact and lose tree-shaking (I think?) or consume the original ES6 source and have to spend build time minifying it myself.  I get around this today by building a webpack DLL once, but again that means I lose tree-shaking.  If I could consume minified ESM then tree-shaking would still work but I wouldn't have to minify myself on every build.
* I was surprised to see you didn't take this opportunity to get rid of `ThirdParty` completely. Just curious, what's the advantage of including these libraries in your distro?
* Have you tried enabling sourcemaps in rollup?  Last time I had to set this up it was just a matter of toggling a single setting in the config file.  I'll keep an eye out for that issue you mentioned.

I'm about to go on holiday but if you're still working on this when I get back, I'll try consuming this version and report on my experience.

mramato commented:
First off, thanks for the feedback @puckey and @thw0rted, nothing here is set in stone and ultimately this is an incremental change designed to get us to ES6 and put us in good shape for further enhanmcenets down the road. The reason we're waiting until after Oct 1st to merge is so that it has a whole month to stew and improve before the Nov 1st release, so keep your thoughts coming throughout the month.

I'll try and answer your questions to the best of my ability, but I'm by no means an expert of ES6 so we're completely willing to change things up if there is good technical/objective reasoning behind it. If you any other links to help educate me on current best practices they would be greatly appreciated.

> * I don't see a lot of `export default` anymore.

It was easier to write a script that used default exports, but now that everything is working (and if there is a good reason to change things) then I'm not against switching.  However, my current thinking is that `export default` being used internally by Cesium's modules is an implementation detail. `Source/Cesium.js`, which I considered the ES6 entry point, uses named exports for all of Cesium's API.  So what would the benefit be to switching the individual files to provide? Do we think users are going to want to include individual ES files directly instead?  Since it's probably only an hour or two's worth of work, we probably don't need to over think it, I just want to develop a better understanding first.  I'll check out @puckey's link and do a little research, but if anyone has any other material on this, that would be great.

> Have you considered distributing a built esm.js artifact?

I'm not sure what you mean here.  Are you talking about a ES6 module where everything is combined and minified?  I tried that but treeshaking didn't work with it, perhaps I'm missing something. If there is a way to combine/minfify into a single ES6 JS file that still allowed for treeshaking, then I would consider not shipping the Source directory at all.

> * I was surprised to see you didn't take this opportunity to get rid of ThirdParty completely. Just curious, what's the advantage of including these libraries in your distro?

One thing at a time. It would have added even more work to start using ThirdParty libraries directly from npm and there would have zero impact on our users. This is definitely a change I want to make, but probably as 1 PR per library since some of them would be non-trivial updates.

> * Have you tried enabling sourcemaps in rollup? Last time I had to set this up it was just a matter of toggling a single setting in the config file. I'll keep an eye out for that issue you mentioned.

I have every intention of getting sourcemaps support in, and I suspect it will be trivial.  I just haven't had a chance to look into it yet and educate myself on them in general. Will probably happen in a separate PR (and in time for the Nov 1st release)

> I'm about to go on holiday 

Hope it's somewhere fun. Even if we do merge this PR before you get back to it, please don't hesitate to provide additional feedback and thoughts.

Thanks again!

hpinkos commented on Source/DataSources/KmlDataSource.js:
```diff
@@ -140,9 +140,8 @@ import WallGraphics from './WallGraphics.js';
-        twitter : false,
         email : false,
-        replaceFn : function(linker, match) {
+        replaceFn : function(match) {
```
Is this because we updated Autolinker?

mramato replied:
Yes

TJKoury commented:
@mramato The `export default` issue is concerning IDE's being able to pick up property / method names from re-exported objects.

Try this in VSCode:

foo.js

```
export const foo = {
  bar: 'bar',
  baz: 'baz',
};
```

masterIndex.js

```
export { foo } from './foo.js';
```
myProject.js

```
import {  foo } from './index.js';
```

Then at the bottom of 'myProject.js' type `foo.`.  Intellisense should give you `bar` and `baz`.

Now go back and change ` export { foo } from './foo.js'; `  to `export {default as foo} from './foo.js`.

Intellisense will not be able to pick up `bar` and `baz` since you're passing an object containing foo, not the original foo module.

TJKoury commented:
I will say it's not just an IDE issue, but it's more apparent there since there isn't really any performance difference or scoping issues (that I've seen).

thw0rted commented:
1.) The link about `export default` has some good points, but I don't think it's black and white.  I personally like named exports because I don't always structure my code as one-file-per-thing.  It's especially relevant in TypeScript because I might export one class from the file, but also a few related constants and maybe some interface definitions for options-objects, etc.  If I'm in the habit of never using `export default`, I just assume I need to do named imports and never wind up with `import Foo, {FooConstructorOptions} from "./foo"`.  I do it for consistency, but it's not a hard rule.  Also, per TJ's comments about IDEs, with named exports I can just start typing the name of something that's in my workspace but not imported yet, and the IDE will notice and make the import for me, which is nice.

2.) As for the ESM rollup output, I'm new to writing ES6 libraries too so take what I'm asking with a grain of salt.  At the very least, I'm sure that what's in this branch now will work better than what I have today.  I'll try to chip in here as I learn more about best practices for this.

3.) I asked about the ThirdParty thing because I encountered a problem a few months back that wound up being a known issue in one of your dependencies (zip.js) that was resolved years prior (see #8092).  I figure if you use NPM to manage dependencies instead of doing it manually, it's easier to keep current, and thus this sort of thing is less likely to happen.

TJKoury commented:
@puckey I just checked and this implementation will work fine in VSCode Intellisense.  I think they must have updated the logic recently.

@thw0rted I removed those comments due to checking ^^^.

![image](https://user-images.githubusercontent.com/749096/65696854-f9d07980-e047-11e9-9f43-666527a8400c.png)

TJKoury commented:
Another note, NodeJS requires [either .mjs extension or package.json with type:'module' ](https://nodejs.org/api/esm.html#esm_enabling).

I'm trying to test a few things using the --experimental-modules flag and I ran across this.

thw0rted commented:
Matt said in his exposition that they're using `esm` to load modules in Node.  I haven't had to yet but it was my impression that you can load `.esm.js` or just `.js` that happens to be a module, using `esm`.  I don't currently consume Cesium from Node but might in the future, so I'll follow this part with interest.

TJKoury commented:
@thw0rted I tried esm but there seems to be some issue with Node v12.x.  @mramato What version of node did you test against?

TJKoury commented:
It's entirely possible it is my own code that is causing the issue, I did not bother isolating it, just converted everything over to .mjs.

mramato commented:
> Another note, NodeJS requires [either .mjs extension or package.json with type:'module' ](https://nodejs.org/api/esm.html#esm_enabling).

Right now nothing should change for Node from an external perspective. You should just be able to require it in as usual.  I tested this with both 10, 12, and 12 with --experimental-modules.  I assume node has support for "legacy" commonjs modules, and that's what we're still providing. Once Node finally brings modules out of experimental, I don't think it would be hard to support them, but probably a future PR.

That being said, I did just find an issue with `NODE_ENV=production` that's on our end, but will hopefully be an easy fix.

TJKoury commented:
@mramato Sorry, I meant the individual modules, requiring the cjs final bundle works fine.  No problem either way, but maybe consider `type:module` as a fairly pain-free way to facilitate.

BTW, ` ko = module.exports;`  in `knockout-3.5.0.js` causes issues, not that this really matters in Node but something I saw while doing tests.

mramato commented:
> BTW, ` ko = module.exports;` in `knockout-3.5.0.js` causes issues, not that this really matters in Node but something I saw while doing tests.

Yes, that's what I'm fixing. It actual does matter because it creates problems when we NODE_ENV=production and we use the optimized version of Cesium for performance.

TJKoury commented:
@mramato Two other things I had to change to be able to pull it up in Node without modifications, checking if `ko` is defined in `knockout-es5.js` and `knockout` is defined in `SvgPathBindingHandler.js`.

mramato commented:
@TJKoury I just pushed a fix. I had to tweak the manual shim around protobuf and knockout to avoid polluting the global namespace and module.exports in Node.  It's these types of problems that make using the third party libraries straight from npm a problem.

TJKoury commented:
@mramato works great.  Any objection to putting `"type":"module"` in the package.json?

mramato commented:
No objections, just need to think through and understand the implications and test it all our (and perhaps generate a separate index.mjs).

kring commented on .gitignore:
```diff
@@ -14,12 +14,12 @@ Thumbs.db
 /Specs/SpecList.js
+/Source/Shaders/**/*.js
+/Source/ThirdParty/Shaders/**/*.js
+/Source/Workers
```
```suggestion
/Source/Workers/**
!/Source/Workers
```

.gitignore can't include files in a directory that has been excluded completely. So the `!/Source/Workers/cesiumWorkerBootstrapper.js` doesn't actually work without my suggested change above.

mramato replied:
@kring is correct here, (nice catch).

@OmarShehata what you are seeing is that git has different behavior if you are ignoring a file that is already submitted to the repository.  If you were to create a new file called Workers/foo.js and then tried to unignore it without Kevin's change, it doesn't work.

kring commented:
Got this working in Terria without any major difficulties. Good work @mramato!

Terria is a mix of CommonJS and ES6 modules at the moment, so I did have to change all the requires in the CommonJS modules from `require('whatever/cesium/module')` to `require('whatever/cesium/module').default`. I guess maybe `esm` papers over that difference in a proper node context?

Speaking of which, if you haven't already, it might be handy to publish this branch to npm under a `beta` tag (i.e. set your version in package.json to `1.62.0-beta-1` or whatever and run `npm publish --tag beta`). That way it can be tested with a proper install from npm.

TJKoury commented:
Recommend in the built shader files using template literals and removing the `\n` artifacts:
```
//This file is automatically rebuilt by the Cesium build process.
export default  `
#ifdef QUANTIZATION_BITS12 
attribute vec4 compressed0; 
attribute float compressed1; 
...
```

vs.

```
//This file is automatically rebuilt by the Cesium build process.
export default "#ifdef QUANTIZATION_BITS12\n\
attribute vec4 compressed0;\n\
attribute float compressed1;\n\
```

mramato commented:
> ecommend in the built shader files using template literals and removing the `\n` artifacts:

Outside the scope of this PR and won't work on IE11.  Remember, we don't transpile down code other than import/export (at least not yet).

mramato commented:
> That way it can be tested with a proper install from npm.

I was thinking of doing this right after we merge up so we basically have the next release + this change with nothing else.  You can also just stick `http://cesium-dev.s3-website-us-east-1.amazonaws.com/cesium/es6-at-last/cesium-1.61.0-es6-at-last32321.tgz` in package.json for the same thing (but obviously doesn't come from npm servers).

mramato commented:
> Terria is a mix of CommonJS and ES6 modules at the moment, so I did have to change all the requires in the CommonJS modules from `require('whatever/cesium/module')` to `require('whatever/cesium/module').default`. I guess maybe `esm` papers over that difference in a proper node context?

@kring This sounds more like a quirk of whatever loader you are using more than anything else.  What system are you using?  I wonder if switching to named modules would have any affect or if it would simple require `require('whatever/cesium/module').Cartesian3` instead (which would actually be worse in my opinion).

markerikson commented:
@mramato , @kring : that's pretty standard behavior any time you're mixing import behavior between different module formats (CommonJS, AMD, ESM).  Or at least, it's the way Webpack (and presumably other loaders) implement it.

mramato commented:
Now that 1.61 is out, I would like to get this merged up to master ASAP.  So if anyone has any critical feedback or strong opinions about changes that affect the end user, please let us know sooner rather than later.  Once this is in master we will absolutely continue to improve it, I'm just trying to avoid any more huge sweeping changes.

@OmarShehata, thanks for volunteering to do a formal review here. Can that happen soon?  I really wanted to this in master like today, but by the end of the week is probably more realistic.

TJKoury commented:
I am using this branch in my celestrak.com build; no issues to report.

mramato commented:
Awesome, thanks @TJKoury!

mramato commented:
While doing some self review, I realized that Cesium.js has some subtle difference (which don't matter in practice) from master.

* Shaders are put at the top level instead of a sub-object, but they were always private to begin with.  So instead of `Cesium._shaders.AcesTonemappingStage` it's not `Cesium._shadersAcesTonemappingStage` I think this is fine since they are still private and it makes the code similar.
* Third party libraries that had special characters in names, such as `Cesium['knockout-3.5.0']` are now exported as `Cesium.knockout_3_5_0`  This is because ES6 restricts export names to valid identifies.

None of these issues are part of the official API, just something I noticed and wanted to bring up.

Currently the CombinedUnminified version in this branch is bigger than master and I have no idea why (it's significantly less lines of code)  Trying to figure out the difference.

OmarShehata commented:
I'm planning on taking a look and doing some testing tomorrow.

TJKoury commented:
@mramato I ran into the ._shaders issue as well; in my working copy I had put all those in a _shaders object that was exported, but since nothing in Cesium uses those in that nomenclature as you mentioned.

Happy to continue testing this branch in production.  Thanks again for cranking this out so quick.

OmarShehata commented on Source/ThirdParty/topojson.js:
```diff
@@ -1,8 +1,8 @@
 /**
 @license
-topojson - https://github.com/mbostock/topojson
+topojson - https://github.com/topojson/topojson
```
Is this diff just updating to the latest topojson version?

mramato replied:
Yes, topojson and Autolinker were both updated, see my original PR description

OmarShehata commented on Specs/Core/RuntimeErrorSpec.js:
```diff
@@ -19,13 +19,22 @@ describe('Core/RuntimeError', function() {
         expect(str).toContain(name + ': ' + testMessage);
-        expect(str).toContain('Core/RuntimeErrorSpec.js');
+
+        if (window.specsUsingRelease) {
```
From the perspective of me writing tests, do I need to be aware of `specsUsingRelease` ? Is it worth documenting before merging?

mramato replied:
This isn't a new flag, it's been there forever and its a very rare occurrence anyone would have to worry about it.  These particular tests are wonky edge cases. The idea is to not be aware of a difference so that we run the exact same specs in dev as release.

OmarShehata commented on Specs/Core/TaskProcessorSpec.js:
```diff
@@ -3,29 +3,24 @@ import { TaskProcessor } from '../../Source/Cesium.js';
 
     it('works with a simple worker', function() {
-        taskProcessor = new TaskProcessor('returnParameters');
+        taskProcessor = new TaskProcessor('returnParameters.js');
```
Why do these need the `.js`? The actual source code calls don't seem to need this?

OmarShehata commented on Specs/TestWorkers/throwError.js:
```diff
@@ -1,12 +1,10 @@
 
     return createTaskProcessorWorker(function(parameters, transferableObjects) {
-        throw new RuntimeError(parameters.message);
+        throw new Error(parameters.message);
```
Why no longer using `RuntimeError` here?

mramato replied:
Because these test-only workers don't go through any kind of build process and `Core/RuntimeError` won't be available to them as part of the built Cesium.js tests.  There was no reason for them to use RuntimeError so it was easiest to just remove it.

OmarShehata commented on Tools/rollup-plugin-strip-pragma/package.json:
```diff
@@ -0,0 +1,25 @@
+  "name": "rollup-plugin-strip-pragma",
+  "version": "1.0.0",
+  "description": "A Rollup plugin to strip requirejs build pragmas from your code.",
+  "homepage": "http://cesiumjs.org",
```
Change to cesium.com

mramato replied:
The homepage for open source CesiumJS related things is still ceisumjs.org, will get changed in the future if/when we do any kind of switchover.

OmarShehata commented on gulpfile.js:
```diff
@@ -25,11 +26,15 @@ var gulpRename = require('gulp-rename');
 var AWS = require('aws-sdk');
 var mime = require('mime');
+var rollup = require('rollup');
+var rollupPluginStripPragma = require('./Tools/rollup-plugin-strip-pragma');
```
Since this is a dev-dependency, and it's now published on npm, we should be able to remove `Tools/` here, right?

mramato replied:
Yes, good catch.

OmarShehata commented on gulpfile.js:
```diff
@@ -131,24 +151,81 @@ gulp.task('buildApps', function() {
+                    '../Source/Cesium.js': 'Cesium',
+                    '../../Source/Cesium.js': 'Cesium',
+                    '../../../Source/Cesium.js': 'Cesium',
+                    '../../../../Source/Cesium.js': 'Cesium'
```
Can we just declare this object once at the top of this function and re-use it, since it seems to be copied 3 times here.

OmarShehata replied:
Also, this doesn't seem like it's supposed to be in the `build-specs` function?

```
var cesiumViewerOutputDirectory = 'Build/Apps/CesiumViewer';
    mkdirp.sync(cesiumViewerOutputDirectory);
```

mramato replied:
Thanks, means to do some self-review of this file and I guess I missed them.  Breaking up gulpfile.js and doing a big clean-up is pretty high on my list of side projects after this is merged.

OmarShehata commented on gulpfile.js:
```diff
@@ -256,7 +296,7 @@ function combine() {
 }
 
-gulp.task('combine', gulp.series('generateStubs', combine));
+gulp.task('combine', gulp.series('build', combine));
```
I never really used `generateStubs`, which looks to be gone now. Is this the WebGL-stubbed version of the tests? Isn't this what travis uses when running the tests?

mramato replied:
No, that's `getWebGLStub`, which is still very much around.  `generateStubs` is what made testing combined/minified CesiumJS possible.  They were requirejs compatible shims that called build version under the hood and were injected into the tests if release/combined was requested.  It was just completely under the hood so you didn't know you were using them, you never had a reason to call it directly.

mramato commented on .eslintrc.json:
```diff
@@ -5,5 +5,18 @@
+                "Source/Workers/transferTypedArrayTest.js"
+            ],
+            "parserOptions": {
+                "sourceType": "script"
```
1. eslint's default is script, but `Tools/eslint-config-cesium/browser.js` is the standard Cesium ecosystem package for browser based development, which now sets `module` to the default. The top-level .eslintrc.json inherits from it.
2. We need to set it back here for `transferTypedArrayTest` because it is not a module, it doesn't get transpiled which is why it lives in Workers and not WorkersES6.  It is used only by the worker system to detect transferable typed array support.
3. cesiumWorkerBootstrapper.js is ignored completely by eslint in master as well, it contains an entire copy of requirejs, verbatim, used by the worker system so linting it isn't useful.

mramato commented on Apps/Sandcastle/load-cesium-es6.js:
```diff
@@ -0,0 +1,10 @@
+// Since ES6 modues have no gauranteed load order,
+// only call startup if it's already defined but hasn't been called yet
+if (!window.startupCalled && typeof window.startup === 'function') {
+    window.startup(Cesium);
```
Fixed

mramato commented:
> * You mentioned writing an issue to generate source maps to get better stack traces for minified Cesium. Did you write that up? If not please do, that'll help a lot when debugging people's issues.

#8245 

> * Should we add "This file is generated" to the Workers/* just like we do with shaders? I can imagine a developer accidentally modifying that file instead of WorkersES6.

Not sure how helpful it will be, but it was easy enough to add so I did it.

> * So ESM is a legit npm dependency of Cesium that gets bundled and ships with the library? So this opens up a path to getting rid of ThirdParty/ and having those all as npm modules? If so, can you write up an issue for that?

No, ESM is just how we load the unbuilt version of Cesium into Node.JS which is makes node-based development easier to debug.  It's not even required if we just always wanted to use the combined version there. It is not bundled or shipped with anything.  This is no different than how we used to depend on requirejs on the node side.

Switching to npm dependencies for Cesium's Source/ThirdParty is not trivial or something we will probably do anytime soon.  The fact that we need to support both Node and the Browser as well as shim some of the libraries makes it tough.  There's also stuff in `Source/ThirdParty` that isn't even on npm.

> * I am leaning towards not using `export default`, just because of this explicit naming philosophy, that you know for sure no one can do `import blah as 'Cartesian3'` and confusing others on the team, but I don't know if this has any adverse consequences.

Nothing stops people from doing this even with named exports, the syntax is just slightly different.  In fact, if we did this then everything we import CesiumMath would be a special case because it would be `Math as CesiumMath`.  If someone can provide a good technical reason for avoid default exports, I'm all ears, but so far it seems like an opinion thing and it's just not worth the time to do the change.  I'm not even against it, it is just another sweeping change that doesn't seem to have rhyme or reason (or time investment) to make.

> * I haven't yet tried how much easier this makes running Cesium in a new webpack/parcel project, but I want to try that and report here.

Any changes to https://cesium.com/docs/tutorials/cesium-and-webpack/ and a PR into https://github.com/AnalyticalGraphicsInc/cesium-webpack-example would be greatly appreciated.

TJKoury commented:
@mramato @OmarShehata Not trying to be pushy, but as I mentioned above adding "type":"module" allows all of the converted modules in the subfolder to be used natively in node with the `--experimental-modules` flag.

No need for `esm`, no side effects.

mramato commented:
@TJKoury it's on my list, no worries.

OmarShehata commented:
I think this is all :+1: from me. @mramato if you have time to look into why standalone mode isn't working here that'd be great, but I'm also willing to merge this and have a priority-next-release issue to fix that. That way it can get more testing in master sooner than later.

mramato commented:
@OmarShehata I'm debugging the standalone thing now. It only affects built sandcastle, but could be the sign of a larger issue when trying to auto-detect CESIUM_BASE_URL. (also have some other changes from your review to push).  I'll bump when ready

mramato commented:
@TJKoury I did an initial attempt of adding `"type": "module" to package.json, but it didn't work. If you can write up an issue (or open a PR) with more context we'll happily take a look at it in time for Nov 1.  We just need any solution that supports Node 10 and 12 both ES6 and commonjs.

mramato commented:
@OmarShehata I believe I addressed all of your issues or otherwise replied to your feedback.  Thanks for the review and catching a few things I missed.

This should be ready.

TJKoury commented:
@mramato make sure to use `node --experimental-modules` to load ES6 modules.

TJKoury commented:
If you still are having issues I'll open an issue.

mramato commented:
> If you still are having issues I'll open an issue.

Yes, please do.  I haven't done anything with ES6 modules in Node yet so I'm sure it's something simple I'm missing.

OmarShehata commented:
:crossed_fingers:

mramato commented:
There's now a blog with more details about this transition up on https://cesium.com/blog/2019/10/31/cesiumjs-es6/
